src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,480 @@
+{
+    double const nu = mean_motion(mu, -semiMajorAxis);
+    double const Mt = M0 + timeSinceEpoch * nu;
+    return solve_kepler_hyperbolic(Mt, eccentricity, 20000);
+}
+
+/**
+ * @brief Compute the position and velocity vectors of an orbiting body at the given anomaly
+ */
+static void get_state_vectors(double   const eccentricAnomaly, 
+                              double   const trueAnomaly, 
+                              double   const eccentricity, 
+                              double   const semiMajorAxis, 
+                              double   const gravitationalParameter, 
+                              Vector3d const u, 
+                              Vector3d const v, 
+                              Vector3d const w,
+                              Vector3d      &position, 
+                              Vector3d      &velocity)
+{
+    double const r = get_orbiting_radius(eccentricity, semiMajorAxis, trueAnomaly);
+    position = r * (cos(trueAnomaly) * u + sin(trueAnomaly) * v);
+    if (eccentricity < 1.0)
+    {
+        velocity = (std::sqrt(gravitationalParameter * semiMajorAxis) / r) * 
+            (-sin(eccentricAnomaly) * u + std::sqrt(1 - eccentricity * eccentricity) * cos(eccentricAnomaly) * v);
+    }
+    else
+    {
+        double angularMomentum = std::sqrt(gravitationalParameter * -semiMajorAxis * (eccentricity * eccentricity - 1));
+
+        double const v_r = (gravitationalParameter / angularMomentum * eccentricity * sin(trueAnomaly));
+        double const v_perp = (angularMomentum / r);
+        Vector3d const radusDir = position.normalized();
+        Vector3d const perpDir = cross(w, radusDir).normalized();
+
+        velocity = v_r * radusDir + v_perp * perpDir;
+    }
+}
+
+} // namespace
+
+
+
+namespace osp
+{
+
+double KeplerOrbit::get_true_anomaly(double const time) const
+{
+    if (is_elliptic())
+    {
+        double const E = time_to_elliptic_eccentric_anomaly(m_params.meanAnomalyAtEpoch, m_params.eccentricity, 
+                    m_params.semiMajorAxis, m_params.gravitationalParameter, time - m_params.epoch);
+        return elliptic_eccentric_anomaly_to_true_anomaly(m_params.eccentricity, E);
+    }
+    else
+    {
+        double const F = time_to_hyperbolic_eccentric_anomaly(m_params.meanAnomalyAtEpoch, m_params.eccentricity, 
+                    m_params.semiMajorAxis, m_params.gravitationalParameter, time - m_params.epoch);
+        return hyperbolic_eccentric_anomaly_to_true_anomaly(m_params.eccentricity, F);
+    }
+}
+
+void KeplerOrbit::get_uvw_vectors(Vector3d &u, Vector3d &v, Vector3d &w) const
+{
+    double const a = m_params.argumentOfPeriapsis;
+    double const b = m_params.longitudeOfAscendingNode;
+    double const c = m_params.inclination;
+
+    u = Vector3d(cos(a) * cos(b) - sin(a) * sin(b) * cos(c),
+                 cos(a) * sin(b) + sin(a) * cos(b) * cos(c),
+                 sin(a) * sin(c));
```

Comment:
is it fair to assume that C++ standard library doesn't provide any special math functions that do much / any of this for us? i'm guessing that cos and sin are all we get?